File: collatz_cst_swap_cocycle/audit_swap_cocycle.rb
#!/usr/bin/env ruby
# Independently implemented exact checker for the adjacent-swap cocycle.

require "digest"

Cylinder = Struct.new(:length, :odd_count, :residue, :endpoint, :pow2, :pow3) do
  def extend(bit)
    raise "invalid bit" unless bit == 0 || bit == 1
    if bit == 0
      lift = endpoint & 1
      intermediate = endpoint + pow3 * lift
      raise "even compatibility" unless (intermediate & 1).zero?
      next_endpoint = intermediate / 2
      next_odd_count = odd_count
      next_pow3 = pow3
    else
      lift = (1 - endpoint) & 1
      intermediate = endpoint + pow3 * lift
      raise "odd compatibility" if (intermediate & 1).zero?
      next_endpoint = (3 * intermediate + 1) / 2
      next_odd_count = odd_count + 1
      next_pow3 = 3 * pow3
    end
    Cylinder.new(length + 1, next_odd_count, residue + pow2 * lift,
                 next_endpoint, 2 * pow2, next_pow3)
  end

  def numerator
    pow2 * endpoint - pow3 * residue
  end

  def margin
    residue - endpoint
  end
end

def inverse_mod(value, modulus)
  old_r, r = value, modulus
  old_s, s = 1, 0
  until r.zero?
    quotient = old_r / r
    old_r, r = r, old_r - quotient * r
    old_s, s = s, old_s - quotient * s
  end
  raise "noninvertible" unless old_r == 1
  old_s % modulus
end

def popcount(value)
  value.to_s(2).count("1")
end

def first_crossings(max_length)
  groups = Hash.new { |hash, key| hash[key] = {} }
  stack = [[0, Cylinder.new(0, 0, 0, 0, 1, 1)]]
  until stack.empty?
    bits, state = stack.pop
    next if state.length == max_length
    [0, 1].each do |bit|
      extension = state.extend(bit)
      extension_bits = bits | (bit << state.length)
      if extension.pow3 < extension.pow2
        groups[extension.length][extension_bits] = extension
      else
        stack << [extension_bits, extension]
      end
    end
  end
  groups
end

def coefficient_gap_coordinates(state)
  gap = state.pow2 - state.pow3
  raise "noncontracting" unless gap.positive?
  margin_residue = if gap == 1
                     0
                   else
                     (-state.numerator * inverse_mod(state.pow3, gap)) % gap
                   end
  window_numerator = state.pow2 * margin_residue + state.numerator
  window_denominator = state.pow2 * gap
  window_index, remainder = window_numerator.divmod(window_denominator)
  raise "window residue" unless remainder == state.residue * gap
  raise "window margin" unless state.margin == margin_residue - gap * window_index
  [gap, margin_residue, window_index]
end

def audit(max_length)
  groups = first_crossings(max_length)
  digest = Digest::SHA256.new
  cylinders = groups.values.sum(&:length)
  edges = 0
  wrapped = 0
  unwrapped = 0
  minimum_jump = nil
  maximum_jump = 0
  maximum_window_index = 0
  wrap_defect_failures = 0

  groups.keys.sort.each do |length|
    states = groups[length]
    modulus = 1 << length
    gap_coordinates = states.transform_values do |state|
      coefficient_gap_coordinates(state)
    end
    maximum_window_index = [maximum_window_index,
                            gap_coordinates.values.map(&:last).max].max
    states.keys.sort.each do |bits|
      source = states.fetch(bits)
      (0...(length - 1)).each do |position|
        next unless ((bits >> position) & 3) == 2
        target = states.fetch(bits ^ (3 << position))
        prefix_ones = popcount(bits & ((1 << position) - 1))
        suffix_ones = popcount(bits >> (position + 2))
        local_modulus = 1 << (length - position)
        inverse = inverse_mod(3**(prefix_ones + 1), local_modulus)
        complement = local_modulus - inverse
        gap = modulus - source.pow3
        suffix_power = 3**suffix_ones
        numerator_delta = (1 << position) * suffix_power
        residue_delta = (1 << position) * inverse
        raise "numerator" unless source.numerator - target.numerator == numerator_delta
        raise "residue" unless target.residue == (source.residue + residue_delta) % modulus

        jump_numerator = gap * inverse + suffix_power
        raise "integrality" unless (jump_numerator % local_modulus).zero?
        positive_jump = jump_numerator / local_modulus
        wrapped_here = source.residue + residue_delta >= modulus
        expected = positive_jump - (wrapped_here ? gap : 0)
        raise "cocycle" unless target.margin - source.margin == expected
        _, margin_residue, window_index = gap_coordinates.fetch(bits)
        _, target_margin_residue, target_window_index =
          gap_coordinates.fetch(bits ^ (3 << position))
        circle_wrap = margin_residue + positive_jump >= gap
        unless target_margin_residue == (margin_residue + positive_jump) % gap
          raise "coefficient-gap rotation"
        end
        unless target_window_index - window_index ==
               (wrapped_here ? 1 : 0) - (circle_wrap ? 1 : 0)
          wrap_defect_failures += 1
        end
        negative_numerator = gap * complement - suffix_power
        raise "negative integrality" unless (negative_numerator % local_modulus).zero?
        negative_jump = negative_numerator / local_modulus
        raise "complement" unless positive_jump + negative_jump == gap
        raise "bounds" unless positive_jump.positive? && negative_jump.positive?

        wrapped_here ? wrapped += 1 : unwrapped += 1
        edges += 1
        minimum_jump = [minimum_jump || positive_jump, positive_jump, negative_jump].min
        maximum_jump = [maximum_jump, positive_jump, negative_jump].max
        digest << format("%d,%x,%d,%d,%d,%d,%d,%d,%d,%d\n",
                         length, bits, position, source.residue, source.endpoint,
                         target.residue, target.endpoint, positive_jump,
                         negative_jump, wrapped_here ? 1 : 0)
      end
    end
  end
  {
    max_length: max_length,
    first_crossing_cylinders: cylinders,
    adjacent_edges: edges,
    unwrapped_edges: unwrapped,
    wrapped_edges: wrapped,
    minimum_jump: minimum_jump || 0,
    maximum_jump: maximum_jump,
    maximum_window_index: maximum_window_index,
    wrap_defect_failures: wrap_defect_failures,
    sha256: digest.hexdigest
  }
end

max_length = Integer(ARGV.fetch(0, "20"))
audit(max_length).each { |key, value| puts "#{key}=#{value}" }
puts "status=independent exact adjacent-swap cocycle audit passed"

File: collatz_cst_swap_cocycle/audit_swap_cocycle.py
"""Exact audit of the adjacent-swap cocycle for CST parity cylinders."""

from __future__ import annotations

import argparse
import hashlib
from collections import defaultdict
from dataclasses import dataclass


@dataclass(frozen=True)
class Cylinder:
    """Least nonnegative representative and endpoint of one parity cylinder."""

    length: int
    odd_count: int
    residue: int
    endpoint: int
    pow2: int
    pow3: int

    @staticmethod
    def empty() -> Cylinder:
        return Cylinder(0, 0, 0, 0, 1, 1)

    def extend(self, bit: int) -> Cylinder:
        if bit not in (0, 1):
            raise ValueError("bit must be 0 or 1")
        if bit == 0:
            lift = self.endpoint & 1
            intermediate = self.endpoint + self.pow3 * lift
            if intermediate & 1:
                raise AssertionError("even extension is incompatible")
            endpoint = intermediate // 2
            odd_count = self.odd_count
            pow3 = self.pow3
        else:
            lift = (1 - self.endpoint) & 1
            intermediate = self.endpoint + self.pow3 * lift
            if not intermediate & 1:
                raise AssertionError("odd extension is incompatible")
            endpoint = (3 * intermediate + 1) // 2
            odd_count = self.odd_count + 1
            pow3 = 3 * self.pow3
        return Cylinder(
            self.length + 1,
            odd_count,
            self.residue + self.pow2 * lift,
            endpoint,
            2 * self.pow2,
            pow3,
        )

    @property
    def numerator(self) -> int:
        """The B in T^K(x)=(3^q*x+B)/2^K."""
        return self.pow2 * self.endpoint - self.pow3 * self.residue

    @property
    def margin(self) -> int:
        """Least-residue descent margin r-T^K(r)."""
        return self.residue - self.endpoint


def first_crossing_cylinders(max_length: int) -> dict[int, dict[int, Cylinder]]:
    """Enumerate words whose coefficient first becomes <1 at their last bit.

    A word is encoded chronologically, with its first parity bit in bit 0.
    Once a branch crosses, it is recorded and not extended.
    """
    if max_length < 1:
        raise ValueError("max_length must be positive")
    groups: dict[int, dict[int, Cylinder]] = defaultdict(dict)
    stack = [(0, Cylinder.empty())]
    while stack:
        bits, state = stack.pop()
        if state.length == max_length:
            continue
        for bit in (0, 1):
            extension = state.extend(bit)
            extension_bits = bits | (bit << state.length)
            if extension.pow3 < extension.pow2:
                groups[extension.length][extension_bits] = extension
            else:
                stack.append((extension_bits, extension))
    return dict(groups)


def audit(max_length: int = 26) -> dict[str, int | str]:
    """Verify every admissible 01->10 edge through ``max_length`` exactly."""
    groups = first_crossing_cylinders(max_length)
    digest = hashlib.sha256()
    cylinders = sum(len(group) for group in groups.values())
    edges = 0
    wrapped = 0
    unwrapped = 0
    minimum_jump = None
    maximum_jump = 0

    for length in sorted(groups):
        states = groups[length]
        modulus = 1 << length
        for bits in sorted(states):
            source = states[bits]
            for position in range(length - 1):
                # Source has 01; target moves that 1 one place earlier to 10.
                if ((bits >> position) & 3) != 2:
                    continue
                target_bits = bits ^ (3 << position)
                target = states.get(target_bits)
                if target is None:
                    raise AssertionError("left shift did not preserve first crossing")

                prefix_ones = (bits & ((1 << position) - 1)).bit_count()
                suffix_ones = (bits >> (position + 2)).bit_count()
                if prefix_ones + suffix_ones + 1 != source.odd_count:
                    raise AssertionError("weight decomposition failed")

                local_modulus = 1 << (length - position)
                inverse = pow(3 ** (prefix_ones + 1), -1, local_modulus)
                complement = local_modulus - inverse
                gap = modulus - source.pow3
                suffix_power = 3**suffix_ones
                numerator_delta = (1 << position) * suffix_power
                residue_delta = (1 << position) * inverse

                if source.numerator - target.numerator != numerator_delta:
                    raise AssertionError("affine numerator swap identity failed")
                expected_residue = (source.residue + residue_delta) % modulus
                if target.residue != expected_residue:
                    raise AssertionError("2-adic residue swap identity failed")

                jump_numerator = gap * inverse + suffix_power
                if jump_numerator % local_modulus:
                    raise AssertionError("jump integrality failed")
                positive_jump = jump_numerator // local_modulus
                wrapped_here = source.residue + residue_delta >= modulus
                expected_margin_delta = positive_jump - (gap if wrapped_here else 0)
                actual_margin_delta = target.margin - source.margin
                if actual_margin_delta != expected_margin_delta:
                    raise AssertionError("margin cocycle identity failed")

                negative_jump_numerator = gap * complement - suffix_power
                if negative_jump_numerator % local_modulus:
                    raise AssertionError("complementary jump integrality failed")
                negative_jump = negative_jump_numerator // local_modulus
                if positive_jump + negative_jump != gap:
                    raise AssertionError("jump complementarity failed")
                if not (0 < positive_jump < gap and 0 < negative_jump < gap):
                    raise AssertionError("first-crossing jump bounds failed")
                if wrapped_here:
                    wrapped += 1
                    if actual_margin_delta != -negative_jump:
                        raise AssertionError("wrapped edge did not decrease margin")
                else:
                    unwrapped += 1
                    if actual_margin_delta != positive_jump:
                        raise AssertionError("unwrapped edge did not increase margin")

                edges += 1
                minimum_jump = (
                    positive_jump
                    if minimum_jump is None
                    else min(minimum_jump, positive_jump, negative_jump)
                )
                maximum_jump = max(maximum_jump, positive_jump, negative_jump)
                digest.update(
                    (
                        f"{length},{bits:x},{position},{source.residue},"
                        f"{source.endpoint},{target.residue},{target.endpoint},"
                        f"{positive_jump},{negative_jump},{int(wrapped_here)}\n"
                    ).encode("ascii")
                )

    return {
        "max_length": max_length,
        "first_crossing_cylinders": cylinders,
        "adjacent_edges": edges,
        "unwrapped_edges": unwrapped,
        "wrapped_edges": wrapped,
        "minimum_jump": minimum_jump or 0,
        "maximum_jump": maximum_jump,
        "sha256": digest.hexdigest(),
    }


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--max-length", type=int, default=26)
    args = parser.parse_args()
    for key, value in audit(args.max_length).items():
        print(f"{key}={value}")
    print("status=exact adjacent-swap cocycle audit passed")


if __name__ == "__main__":
    main()
